src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -9247,8 +9252,19 @@ std::pair<bool, NameSet> StorageReplicatedMergeTree::unlockSharedDataByID(
         }
         else
         {
-            LOG_TRACE(logger, "Can't remove parent zookeeper lock {} for part {}, because children {} ({}) exists",
-                zookeeper_part_node, part_name, children.size(), fmt::join(children, ", "));
+            /// It's possible that we have two instances of the same part with different blob names of
+            /// FILE_FOR_REFERENCES_CHECK aka checksums.txt aka part_unique_id,
+            /// and other files in both parts are hardlinks (the same blobs are shared between part instances).
+            /// It's possible after unsuccessful attempts to commit a mutated part to zk.
+            /// It's not a problem if we have found the mutation parent (so we have files_not_to_remove).
+            /// But in rare cases mutations parents could have been already removed (so we don't have the list of hardlinks).
+
+            /// I'm not 100% sure that parent_not_to_remove list cannot be incomplete (when it's not empty)
+            if (part_info.mutation && parent_not_to_remove.empty())
```

Comment:
I see the comment, but still doesn't understand, sorry. 

I expected to see here some straightforward logic like check in zookeeper our "sibling" (this another part with different "FILE_FOR_REFERENCES_CHECK aka checksums.txt aka part_unique_id") hardlinks.

Minor: Why not just check `has_parent` flag? Why we explicitly check list of `parent_not_to_remove`?

Reply:
The problem is that we don't have the list of hardlinks in the sibling part. We had it only in the parent part. And we have a problem if the parent part is already removed (we don't know the list of hardlinks in this case, so we have too keep all blobs)

> Minor: Why not just check has_parent flag?

Yes, looks like it's better to replace `parent_not_to_remove.empty()` with `!has_parent`